#!/usr/bin/env python3
# A model for the UHGG collection of genomes (aka UHGG database).
import os
from collections import defaultdict
from iggtools.params.outputs import genomes as TABLE_OF_CONTENTS
from iggtools.common.utils import select_from_tsv, sorted_dict, InputStream, download_reference, multithreading_map, command, num_physical_cores
from iggtools.params import outputs, inputs
from iggtools.params.inputs import igg


MARKER_FILE_EXTS = ["fa", "fa.bwt", "fa.header", "fa.sa", "fa.sequence", "map"]

def get_uhgg_layout(species_id, component="", genome_id=""):
    return {
        "genomes_toc":                f"genomes.tsv",

        # Used by midas_run_species and generated by collate_repgenome_markers
        "marker_db":                  f"marker_genes/{inputs.marker_set}/{inputs.marker_set}.{component}",
        "marker_db_hmm_cutoffs":      f"marker_genes_models/{inputs.marker_set}/marker_genes.mapping_cutoffs{component}",
        "marker_centroids":           f"marker_genes/{inputs.marker_set}/marker_centroids/{species_id}.txt",
        "marker_centroids_log":       f"marker_genes/{inputs.marker_set}/marker_centroids.log",

        "marker_centroid_99":           f"marker_genes/{inputs.marker_set}/marker_centroids_99/{species_id}.txt",
        "marker_centroids_99_log":       f"marker_genes/{inputs.marker_set}/marker_centroids_99.log",

        # marker_genes/phyeco/temp/{SPECIES_ID}/{GENOME_ID}/{GENOME_ID}.{hmmsearch, markers.fa, markers.map}
        "marker_collate_log":         f"marker_genes/{inputs.marker_set}/collate_repgenome_markers.log",
        "marker_genes":               f"marker_genes/{inputs.marker_set}/temp/{species_id}/{genome_id}/{genome_id}.{component}",

        # gene_annotations/{SPECIES_ID}/{GENOME_ID}/{GENOME_ID}.{fna, faa, gff, log}
        "annotation_file":            f"gene_annotations/{species_id}/{genome_id}/{genome_id}.{component}",

        "imported_genome_file":       f"cleaned_imports/{species_id}/{genome_id}/{genome_id}.{component}",

        # f"{inputs.uhgg_genomes}/{representative_id}/{genome_id}.fna.lz4"
        #"raw_genome_file":            f"{inputs.uhgg_genomes}/{species_id}/{genome_id}.fna.lz4",
        #"imported_genome_file":       f"{outputs.cleaned_imports}/{species_id}/{genome_id}/{genome_id}.{component}",

        # pangenomes/100001/{genes.ffn, centroids.ffn, gene_info.txt}.lz4
        "pangenome_file":             f"pangenomes/{species_id}/{component}",
    }

### old codes, improve the readibility
## there two functions were used in import_uhgg.py leave it alone for now
def imported_genome_file(genome_id, species_id, component):
    return f"{outputs.cleaned_imports}/{species_id}/{genome_id}/{genome_id}.{component}"
def raw_genome_file(genome_id, representative_id):
    return f"{inputs.uhgg_genomes}/{representative_id}/{genome_id}.fna.lz4"


class MIDAS_IGGDB: # pylint: disable=too-few-public-methods

    def __init__(self, midas_iggdb_dir=".", num_cores=num_physical_cores):
        self.midas_iggdb_dir = midas_iggdb_dir
        self.num_cores = num_cores
        self.local_toc = _fetch_file_from_s3((outputs.genomes, self.get_target_layout("genomes_toc", "", "", "", False)))
        self.uhgg = UHGG(self.local_toc)
        self.igg = igg


    def get_target_layout(self, filename, remote, component="", species_id="", genome_id=""):
        local_path = get_uhgg_layout(species_id, component, genome_id)[filename]
        target_filepath = os.path.join(self.igg, f"{local_path}.lz4") if remote else os.path.join(self.midas_iggdb_dir, local_path)
        return target_filepath


    def fetch_files(self, filetype, list_of_species_ids=""):
        """ Fetch files from s3://microbiome-igg/2.0/ to local midas iggdb """
        args_list = []
        fetched_files = {}

        if filetype == "marker_db":
            for ext in MARKER_FILE_EXTS:
                s3_file = self.get_target_layout("marker_db", remote=True, component=ext)
                dest_file = self.get_target_layout("marker_db", remote=False, component=ext)
                fetched_files[ext] = _fetch_file_from_s3((s3_file, dest_file))
            return fetched_files

        # Download per species files
        if list_of_species_ids:
            for species_id in list_of_species_ids:
                if filetype == "contigs":
                    s3_file = self.get_target_layout("imported_genome_file", True, "fna", species_id, self.uhgg.representatives[species_id])
                    dest_file = self.get_target_layout("imported_genome_file", False, "fna", species_id, self.uhgg.representatives[species_id])
                if filetype == "centroids":
                    s3_file = self.get_target_layout("pangenome_file", True, "centroids.ffn", species_id)
                    dest_file = self.get_target_layout("pangenome_file", False, "centroids.ffn", species_id)
                if filetype == "genes_info":
                    s3_file = self.get_target_layout("pangenome_file", True, "gene_info.txt", species_id)
                    dest_file = self.get_target_layout("pangenome_file", False, "gene_info.txt", species_id)
                if filetype == "marker_genes_fa":
                    s3_file = self.get_target_layout("marker_genes", True, "markers.fa", species_id, self.uhgg.representatives[species_id])
                    dest_file = self.get_target_layout("marker_genes", False, "markers.fa", species_id, self.uhgg.representatives[species_id])
                if filetype == "marker_genes_map":
                    s3_file = self.get_target_layout("marker_genes", True, "markers.map", species_id, self.uhgg.representatives[species_id])
                    dest_file = self.get_target_layout("marker_genes", False, "markers.map", species_id, self.uhgg.representatives[species_id])
                if filetype == "marker_centroids":
                    s3_file = self.get_target_layout("marker_centroids", True, "", species_id)
                    dest_file = self.get_target_layout("marker_centroids", False, "", species_id)
                if filetype == "gene_feature":
                    s3_file = self.get_target_layout("annotation_file", True, "genes", species_id, self.uhgg.representatives[species_id])
                    dest_file = self.get_target_layout("annotation_file", False, "genes", species_id, self.uhgg.representatives[species_id])
                if filetype == "gene_seq":
                    s3_file = self.get_target_layout("annotation_file", True, "ffn", species_id, self.uhgg.representatives[species_id])
                    dest_file = self.get_target_layout("annotation_file", False, "ffn", species_id, self.uhgg.representatives[species_id])
                if filetype == "prokka_genome":
                    s3_file = self.get_target_layout("annotation_file", True, "fna", species_id, self.uhgg.representatives[species_id])
                    dest_file = self.get_target_layout("annotation_file", False, "fna", species_id, self.uhgg.representatives[species_id])
                args_list.append((s3_file, dest_file))

            _fetched_files = multithreading_map(_fetch_file_from_s3, args_list, num_threads=self.num_cores)
            for species_index, species_id in enumerate(list_of_species_ids):
                fetched_files[species_id] = _fetched_files[species_index]
            return fetched_files

        # Fetch single file with filetype as the uhgg layout key
        if filetype in get_uhgg_layout(species_id=""):
            s3_file = self.get_target_layout(filetype, True)
            dest_file = self.get_target_layout(filetype, False)
        else:
            s3_file = os.path.join(self.igg, f"{filetype}.lz4")
            dest_file = os.path.join(self.midas_iggdb_dir, filetype)
        return _fetch_file_from_s3((s3_file, dest_file))


def _fetch_file_from_s3(packed_args):
    s3_path, dest_file = packed_args

    local_dir = os.path.dirname(dest_file)
    if not os.path.isdir(local_dir):
        command(f"mkdir -p {local_dir}")

    if os.path.exists(dest_file):
        return dest_file
    return download_reference(s3_path, local_dir)


class UHGG:  # pylint: disable=too-few-public-methods

    def __init__(self, table_of_contents_tsv=TABLE_OF_CONTENTS):
        # Default is to fetch the UHGG table of contents from S3.
        # However, if already downloaded, the local TOC may be passed as a param.
        self.toc_tsv = table_of_contents_tsv
        self.species, self.representatives, self.genomes = _UHGG_load(table_of_contents_tsv)


    def fetch_representative_genome_id(self, species_id):
        return self.representatives[species_id]


def _UHGG_load(toc_tsv, deep_sort=False):
    species = defaultdict(dict)
    representatives = {}
    genomes = {}
    with InputStream(toc_tsv) as table_of_contents:
        for row in select_from_tsv(table_of_contents, selected_columns=["genome", "species", "representative", "genome_is_representative"]):
            genome_id, species_id, representative_id, _ = row
            species[species_id][genome_id] = row
            representatives[species_id] = representative_id
            genomes[genome_id] = species_id
    if deep_sort:
        for sid in species.keys():
            species[sid] = sorted_dict(species[sid])
        species = sorted_dict(species)
    return species, representatives, genomes
